moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java

```diff
@@ -0,0 +1,193 @@
+      for (int i = 0; i < typeArgs.length; ++i) {
+        var typeVarName = typeVars[i].getName();
+        var materialized = typeArgs[i];
+        mappedTypeArgs.put(typeVarName, materialized);
+      }
+    }
+    var components = rawType.getRecordComponents();
+    var bindings = new LinkedHashMap<String, ComponentBinding<?>>();
+    var constructorParams = new Class<?>[components.length];
+    for (int i = 0, componentsLength = components.length; i < componentsLength; i++) {
+      RecordComponent component = components[i];
+      constructorParams[i] = component.getType();
+      var name = component.getName();
+      var componentType = component.getGenericType();
+      if (componentType instanceof TypeVariable<?> typeVariable) {
+        var typeVarName = typeVariable.getName();
+        if (mappedTypeArgs == null) {
+          throw new AssertionError("No mapped type arguments found for type '" + typeVarName + "'");
+        }
+        var mappedType = mappedTypeArgs.get(typeVarName);
+        if (mappedType == null) {
+          throw new AssertionError("No materialized type argument found for type '" + typeVarName + "'");
+        }
+        componentType = mappedType;
+      }
+      var jsonName = name;
+      Set<Annotation> qualifiers = null;
+      for (var annotation : component.getDeclaredAnnotations()) {
+        if (annotation instanceof Json jsonAnnotation) {
+          jsonName = jsonAnnotation.name();
+        } else {
+          if (annotation.annotationType().isAnnotationPresent(JsonQualifier.class)) {
+            if (qualifiers == null) {
+              qualifiers = new LinkedHashSet<>();
+            }
+            qualifiers.add(annotation);
+          }
+        }
+      }
+      if (qualifiers == null) {
+        qualifiers = Collections.emptySet();
+      }
+      var adapter = moshi.adapter(componentType, qualifiers);
+      var accessor = component.getAccessor();
+      accessor.setAccessible(true);
+      var componentBinding = new ComponentBinding<>(name, jsonName, adapter, accessor);
+      var replaced = bindings.put(jsonName, componentBinding);
+      if (replaced != null) {
+        throw new IllegalArgumentException(
+            "Conflicting components:\n"
+                + "    "
+                + replaced.name
+                + "\n"
+                + "    "
+                + componentBinding.name);
+      }
+    }
+    Constructor<Object> constructor;
+    try {
+      //noinspection unchecked
+      constructor = (Constructor<Object>) rawType.getDeclaredConstructor(constructorParams);
+      constructor.setAccessible(true);
+    } catch (NoSuchMethodException e) {
+      throw new AssertionError(e);
+    }
+    return new RecordJsonAdapter<>(constructor, bindings).nullSafe();
+  };
+
+  private static record ComponentBinding<T>(String name, String jsonName, JsonAdapter<T> adapter, Method accessor) {
+  }
+
+    private final String targetClass;
```

Comment:
Wrong indent. GJF bug?

Reply:
GJF bug, I'll file something over there

Reply:
Actually I think spotless is ignoring this? Let me look more

Reply:
Yeah spotless is ignoring this. Formatting it with GJF manually for now

Reply:
d98ee90